### Numpy introducction
<h3 style="color:#c1528e">Why NumPy?</h3>

##### Numpy es la **columna vertebral del ML**.
##### -Numpy esta escrito en C, lo que lo convierte en un framework super rapido.
##### Vectorization via brodcasting **(avoiding loops)**
##### -Por este motivo se usa en vez de usar *vanilla Python*
### What are we going to cover?
* Most useful functions
* NumPy datatypes and attributes (ndarray)
* Creating arrays
* Manipulating and comparing arrays
* Sorting arrays
* Use cases



 ![Arrays en NumPy](numpyArr.png)

In [1]:
import pandas as pd
import numpy as np

In [20]:
# same array as slide video

a3 = np.array([[[1, 2, 3],
                [4, 5, 6]],
               [[7, 8, 9],
                [10, 11, 12]],
               [[13, 14, 15],
                [16, 17, 18]]])

# view array
print(a3)

[[[ 1  2  3]
  [ 4  5  6]]

 [[ 7  8  9]
  [10 11 12]]

 [[13 14 15]
  [16 17 18]]]


<h1 style="color:yellow"> DataTypes & Attributes</h1>

In [38]:
# NumPy's main datatype is ndarray

arr_one = np.array([0,1,2])
print(type(arr_one))
print("Array 1: ",arr_one)

<class 'numpy.ndarray'>
Array 1:  [0 1 2]


In [39]:
arr_two = np.array([[0, 1, 2],
                   [3, 4, 5]])
print("Array 2:",arr_two)
print("")

# tercer tipo

arr_three = np.array([[[0, 1, 2],
                       [3, 4, 5]],
                       [[6, 7, 8],
                        
                      [9, 10, 11]],
                       [[12, 13, 14],
                       [15, 16, 17]]])
print("Array 3:", arr_three)


Array 2: [[0 1 2]
 [3 4 5]]

Array 3: [[[ 0  1  2]
  [ 3  4  5]]

 [[ 6  7  8]
  [ 9 10 11]]

 [[12 13 14]
  [15 16 17]]]


In [35]:
# shapes of an array

print(arr_one.shape)
print(arr_two.shape)
print(arr_three.shape)

(3,)
(2, 3)
(3, 2, 3)


In [40]:
# ver las dimensiones de los ndarrays

print("Array one # dimensiones:",arr_one.ndim)
print("Array two # dimensiones:",arr_two.ndim)
print("Array three # dimensiones:",arr_three.ndim)

Array one # dimensiones: 1
Array two # dimensiones: 2
Array three # dimensiones: 3


In [43]:
# tipo de dato de los ndarrays

print(arr_one.dtype)
print(arr_two.dtype)
print(arr_three.dtype)
# todos los mioos son enteros, sin embargo se aceptan de cualquier formato.


int32
int32
int32


In [44]:
print(arr_one.size)
print(arr_two.size)
print(arr_three.size)

3
6
18


In [45]:
type(arr_one)

numpy.ndarray

In [46]:
# no matter what type or what shape the numpy array is, all of these are ndarray

In [47]:
print(arr_three)

[[[ 0  1  2]
  [ 3  4  5]]

 [[ 6  7  8]
  [ 9 10 11]]

 [[12 13 14]
  [15 16 17]]]


In [48]:
# Create a df from NumPy Arr
df = pd.DataFrame(arr_two)
print(df)


   0  1  2
0  0  1  2
1  3  4  5


In [52]:

arr_three


array([[[ 0,  1,  2],
        [ 3,  4,  5]],

       [[ 6,  7,  8],
        [ 9, 10, 11]],

       [[12, 13, 14],
        [15, 16, 17]]])

In [54]:
dataFrEx = pd.DataFrame(arr_two) #podemos insertar los array creados en NumPy, como pararmetros del dataframe.
print(dataFrEx)

   0  1  2
0  0  1  2
1  3  4  5


In [3]:
# Creating my own array with NumPy
my_own_array = np.array([[0,1,2],[3,4,5],[6,7,8]])
my_own_df = pd.DataFrame(my_own_array)
print(my_own_df)

print("Size: ", my_own_df.size)
print("Tipo de dato de cada columna", my_own_df.dtypes)
print("Shape: ", my_own_df.shape)
print("Numero de dimensiones: ", my_own_df.ndim)
print("Tipo de objeto de NumPy:",type(my_own_df))




   0  1  2
0  0  1  2
1  3  4  5
2  6  7  8
Size:  9
Tipo de dato de cada columna 0    int32
1    int32
2    int32
dtype: object
Shape:  (3, 3)
Numero de dimensiones:  2
Tipo de objeto de NumPy: <class 'pandas.core.frame.DataFrame'>


## 2-Creating arrays

In [5]:
sample_array = np.array([1,2,3])
sample_array

array([1, 2, 3])

In [7]:
sample_array.dtype

dtype('int32')

In [8]:
ones = np.ones((2,3)) #crea 2 filas (row) y 3 columns
print(ones)

[[1. 1. 1.]
 [1. 1. 1.]]


In [9]:
ones.dtype

dtype('float64')

In [10]:
# creando un array pero que venga relleno de 0
zeros_arr = np.zeros((2,3))
print(zeros_arr)

[[0. 0. 0.]
 [0. 0. 0.]]


#### En ocasiones cuando estemos trabajando con ML vamos a necesitar unicamente calcular las formas de los datos con los que estaremos trabajando, por lo cual usar .zeros() seria de ayuda

In [6]:
# este metodo le indica algo asi como: empezar en 0, parar hasta 10
# y aumentar de 2 en 2, es decir .arange(0,10,2)
range_array = np.arange(0,10,2)
range_array


# creadno mi propio array:
my_own_arange = np.arange(1,11,1)
print("Este es mi propio .arange(arange([start,] stop[, step,], dtype=None, *, like=None)) el cual inicia en 1, llega hasta 10 y incrementa de 1 en 1.")
print(my_own_arange)

Este es mi propio .arange(arange([start,] stop[, step,], dtype=None, *, like=None)) el cual inicia en 1, llega hasta 10 y incrementa de 1 en 1.
[ 1  2  3  4  5  6  7  8  9 10]


In [7]:
# a lo  entendido con los docs, esto devolvera 3 arrays, con tamano 5 cada array,
# con valores de 0 a maximo 10 (9), el size= seria nuestro shape, creo
random_arr = np.random.randint(0,10, size=(3,5))
print(random_arr)

[[5 6 2 2 7]
 [0 7 9 0 4]
 [9 1 7 7 5]]


In [8]:
random_arr.size #cuenta cada dato de los 3 array

15

In [10]:
random_arr.shape


(3, 5)

#### Forma numero 2 de crear random arrys:

In [13]:

# le damos el shape que queremos como parametro, es decir que 
# nos dara un total de 7 rows con 7 columns, pero el valor de cada dato,
# estara entre 0.0 y 1.0, ej: 0.434324

random_arr_two = np.random.random((7,7))
print(random_arr_two)

[[0.54688594 0.578906   0.89362881 0.86061839 0.64485249 0.12611359
  0.3341334 ]
 [0.8979634  0.08697567 0.12297125 0.00798176 0.24340783 0.07221092
  0.69729048]
 [0.98983844 0.5786583  0.33571015 0.35371728 0.87066204 0.00309862
  0.05986099]
 [0.60330119 0.29792002 0.26307607 0.36017599 0.21396139 0.66893243
  0.50830063]
 [0.81957424 0.1503638  0.12646036 0.98558349 0.09720833 0.59432185
  0.21543905]
 [0.28465672 0.15974091 0.51537059 0.10651251 0.92595165 0.74857089
  0.99850659]
 [0.92264554 0.33912329 0.93059937 0.839017   0.38429291 0.10857076
  0.88288547]]


In [14]:
random_arr_two.shape

(7, 7)

### Tercer forma de crear un array random:

numpy.random.rand()

In [15]:
random_arr_three = np.random.rand(7,7)
print(random_arr_three)

[[0.22615361 0.44793414 0.2780255  0.55181322 0.90117194 0.09538584
  0.0247582 ]
 [0.33015585 0.73224435 0.96026668 0.56382692 0.11337934 0.10627365
  0.59304825]
 [0.18572354 0.1777211  0.05577531 0.13351921 0.20318948 0.47241944
  0.81510414]
 [0.60667536 0.9775823  0.02072365 0.18132671 0.48369042 0.61247462
  0.06939612]
 [0.00679627 0.48919021 0.10798141 0.3978099  0.86758194 0.49351665
  0.0042259 ]
 [0.40933013 0.409396   0.46389137 0.46256209 0.64895463 0.64484439
  0.89325129]
 [0.28141985 0.82201473 0.4608619  0.2246404  0.61181329 0.87754469
  0.99861756]]


### Pseudo-random numbers
#### np.random.seed() 

np.random.seed(seed=0)
random_seed = np.random.randint(10, size=(4,3))
print(random_seed)
Output: 
[[8 4 5]
 [3 1 1]
 [1 7 3]
 [0 5 7]]

**Con .seed(seed=0) logramos una secuencia reproducible, es decir que cada vez que ejecutemos, se gurdara el mismo array SIEMPRE.
En cambio si quitamos el .seed() los numeros si van a ser random cada vez que el codigo se ejecute.**

In [24]:
np.random.seed(seed=0)
random_array_four = np.random.randint(10, size=(4,3))
print('Valores random hasta el 9, con 4 filas con 3 columnas:')
print(random_array_four)

Valores random hasta el 9, con 4 filas con 3 columnas:
[[5 0 3]
 [3 7 9]
 [3 5 2]
 [4 7 6]]


Daniel said:

Generalmente se usa cuando trabajamos con algun companero y necesitamos por alguna razon que en las celdas se ejecute **exactamente** el mismo valor de los random numbers.

In [34]:
np.random.seed(seed=1)
unique_array = np.random.randint(11, size=(3,4))
print(unique_array)

[[5 8 9 5]
 [0 0 1 7]
 [6 9 2 4]]


In [25]:
# Como encontrar los numeros unicos en un array?
# np.unique(my_own_array)
# Output: [1,2,3,4]


In [35]:
# buscar los numeros unicos:
my_unique_numbers_array = np.unique(unique_array)
print(my_unique_numbers_array)

[0 1 2 4 5 6 7 8 9]


###  *VIDEO 65*
## Viewing Arrays and Matrices

In [41]:
print('Array 1:',arr_one)
print('')
print('Array 2:',arr_two)
print('')
print('Array 3:',arr_three)


Array 1: [0 1 2]

Array 2: [[0 1 2]
 [3 4 5]]

Array 3: [[[ 0  1  2]
  [ 3  4  5]]

 [[ 6  7  8]
  [ 9 10 11]]

 [[12 13 14]
  [15 16 17]]]


Algunas pracctices para tantear los arreeglos:

In [46]:
arr_one[0]

0

In [54]:
arr_two[1][1]

4

In [56]:
arr_three

array([[[ 0,  1,  2],
        [ 3,  4,  5]],

       [[ 6,  7,  8],
        [ 9, 10, 11]],

       [[12, 13, 14],
        [15, 16, 17]]])

In [66]:
arr_three[2]

array([[12, 13, 14],
       [15, 16, 17]])

Volviendo al video:


In [67]:
print(arr_three)

[[[ 0  1  2]
  [ 3  4  5]]

 [[ 6  7  8]
  [ 9 10 11]]

 [[12 13 14]
  [15 16 17]]]


In [83]:
print('El array 1:',arr_one)
print(arr_one[:1]) #me muestra solo lo que esta antes del indice (no incluye el indice)
print(arr_one[:2])

El array 1: [0 1 2]
[0]
[0 1]


Ejemplos con el arr_two:

In [98]:
print('El array 2:',arr_two)
print('')
print(arr_two[:2,:2])
print('')
print(arr_two[:2,:1])


El array 2: [[0 1 2]
 [3 4 5]]

[[0 1]
 [3 4]]

[[0]
 [3]]


Vamos con el arr_three:

In [100]:
print('El array 3:',arr_three)
print('')
print('Examples')
print(arr_three[:2,:2,:2])

El array 3: [[[ 0  1  2]
  [ 3  4  5]]

 [[ 6  7  8]
  [ 9 10 11]]

 [[12 13 14]
  [15 16 17]]]

[[[ 0  1]
  [ 3  4]]

 [[ 6  7]
  [ 9 10]]]

Examples
[[[0 1 2]]

 [[6 7 8]]]


In [10]:
a7 = np.random.randint(6,8, size=(2,3,4,5))
print(a7)
print('')
# print('El 0:',a7[0])

[[[[7 6 7 7 6]
   [6 6 6 7 6]
   [7 7 7 7 7]
   [7 7 7 7 6]]

  [[6 7 6 7 7]
   [7 6 7 6 6]
   [7 7 7 7 7]
   [6 6 7 7 7]]

  [[6 6 7 6 6]
   [6 7 6 6 7]
   [7 6 7 6 7]
   [6 7 7 7 7]]]


 [[[7 6 7 6 6]
   [7 7 7 6 6]
   [7 7 7 6 6]
   [7 7 7 7 6]]

  [[6 7 7 7 6]
   [6 6 6 7 6]
   [6 7 7 7 7]
   [6 7 7 6 7]]

  [[7 6 7 7 7]
   [7 6 7 6 7]
   [6 7 7 7 6]
   [6 6 7 7 6]]]]



In [8]:
a7.shape, a7.ndim

((2, 3, 4, 5), 4)

In [15]:
# Get the first 3 numbers of the inner most arrays
a7[:,:,:, :3]

array([[[[7, 6, 7],
         [6, 6, 6],
         [7, 7, 7],
         [7, 7, 7]],

        [[6, 7, 6],
         [7, 6, 7],
         [7, 7, 7],
         [6, 6, 7]],

        [[6, 6, 7],
         [6, 7, 6],
         [7, 6, 7],
         [6, 7, 7]]],


       [[[7, 6, 7],
         [7, 7, 7],
         [7, 7, 7],
         [7, 7, 7]],

        [[6, 7, 7],
         [6, 6, 6],
         [6, 7, 7],
         [6, 7, 7]],

        [[7, 6, 7],
         [7, 6, 7],
         [6, 7, 7],
         [6, 6, 7]]]])

### 64. Manipulating & comparing arrays
**Como manipular y comparar nuestras diferentes matrices/arrays**

***Arithmetic***  
*Fancy word for basic math equations*

In [6]:
a1 = np.array([1,2,3])
print(a1)
ones = np.ones(3)
print(ones)
print('')
print(a1 + ones)

# a1(1) + ones(1) = 2.

[1 2 3]
[1. 1. 1.]

[2. 3. 4.]


In [23]:
print(a1 - ones)
# a1(1) - ones(1) = 0

[0. 1. 2.]


In [25]:
print(a1 * ones)
# a1(1) * ones(1) 1 * 1 = 1

[1. 2. 3.]


In [7]:
a2 = np.array([[0, 1, 2.3],
                   [3, 4.7, 5]])
print(a2)

[[0.  1.  2.3]
 [3.  4.7 5. ]]


In [29]:
print(a1 * a2)

[[ 0.   2.   6.9]
 [ 3.   9.4 15. ]]


In [8]:
a3 = np.array([[[0, 1, 2],
                       [3, 4, 5]],
                      [[6, 7, 8],

                      [9, 10, 11]],
                      [[12, 13, 14],
                       [15, 16, 17]]])
print(a3)


[[[ 0  1  2]
  [ 3  4  5]]

 [[ 6  7  8]
  [ 9 10 11]]

 [[12 13 14]
  [15 16 17]]]


In [34]:
print(a3)

[[[ 0  1  2]
  [ 3  4  5]]

 [[ 6  7  8]
  [ 9 10 11]]

 [[12 13 14]
  [15 16 17]]]


In [35]:
print(a2*a3)

[[[ 0.   1.   4.6]
  [ 9.  18.8 25. ]]

 [[ 0.   7.  18.4]
  [27.  47.  55. ]]

 [[ 0.  13.  32.2]
  [45.  75.2 85. ]]]


In [36]:
a2

array([[0. , 1. , 2.3],
       [3. , 4.7, 5. ]])

In [37]:
a3

array([[[ 0,  1,  2],
        [ 3,  4,  5]],

       [[ 6,  7,  8],
        [ 9, 10, 11]],

       [[12, 13, 14],
        [15, 16, 17]]])

In [40]:
a2*a3
a2.shape, a3.shape

((2, 3), (3, 2, 3))


### Array brodcasting:
El término brodcasting describe cómo NumPy trata las matrices con diferentes shapes durante las operaciones aritméticas. Sujeto a ciertas restricciones, la matriz más pequeña se "brodcast" a través de la matriz más grande para que tengan formas compatibles. La transmisión proporciona un medio para vectorizar las operaciones de matriz para que el bucle se produzca en C en lugar de en Python. Lo hace sin hacer copias innecesarias de datos y, por lo general, conduce a implementaciones de algoritmos eficientes. Sin embargo, hay casos en los que la transmisión es una mala idea porque conduce a un uso ineficiente de la memoria que ralentiza el cálculo.



In [41]:
print(a1 / ones)

[1. 2. 3.]


In [42]:
print(a2 / a1)

[[0.         0.5        0.76666667]
 [3.         2.35       1.66666667]]


In [44]:
# floor division remove the decimals ( // )
print(a2 // a1)

[[0. 0. 0.]
 [3. 2. 1.]]


In [46]:
print(a2)
print(a2 ** 2)

[[0.  1.  2.3]
 [3.  4.7 5. ]]
[[ 0.    1.    5.29]
 [ 9.   22.09 25.  ]]


In [47]:
# numPy tiene una funcion para pbtener el cuadrado:
np.square(a2)

array([[ 0.  ,  1.  ,  5.29],
       [ 9.  , 22.09, 25.  ]])

**.square()** solo es apto para obtener el cuadrado,
es decir, que para las demas operaciones aritmeticas: suma, resta, etc, se pueden hacer manuales como en Python o utilizar:  

**.add()**:  Add arguments element-wise.

Parameters:  
**x1, x2**
The arrays to be added.  If x1.shape != x2.shape, they must be broadcastable to a common shape (which becomes the shape of the output).  The sum of x1 and x2, element-wise. This is a scalar if both x1 and x2 are scalars.

*Equivalent to x1 + x2 in terms of array broadcasting.*


In [51]:
# example:
np.add(1.0, 4.0)
5.0
x1 = np.arange(9.0).reshape((3, 3))
x2 = np.arange(3.0)
np.add(x1, x2)

array([[ 0.,  2.,  4.],
       [ 3.,  5.,  7.],
       [ 6.,  8., 10.]])

In [52]:
print(np.add(a2,a3))


[[[ 0.   2.   4.3]
  [ 6.   8.7 10. ]]

 [[ 6.   8.  10.3]
  [12.  14.7 16. ]]

 [[12.  14.  16.3]
  [18.  20.7 22. ]]]


In [55]:
print(a1 % a2)
print(np.mod(a1,a2))
# basicamente es lo mismo, diferentes maneras de hacerlos.

[[nan 0.  0.7]
 [1.  2.  3. ]]
[[nan 0.  0.7]
 [1.  2.  3. ]]


C:\Users\alvis\AppData\Local\Temp\ipykernel_8908\793242999.py:1: RuntimeWarning: invalid value encountered in remainder
  print(a1 % a2)
C:\Users\alvis\AppData\Local\Temp\ipykernel_8908\793242999.py:2: RuntimeWarning: invalid value encountered in remainder
  print(np.mod(a1,a2))


In [3]:
print("Machine Learning Engineer")

Machine Learning Engineer


In [58]:
# np.exp() ->
# Calculate the exponential of all elements in the input array.
print(np.exp(a1))

[ 2.71828183  7.3890561  20.08553692]


In [59]:
a10 = np.array([10,20,30])
print(np.exp(a10))

[2.20264658e+04 4.85165195e+08 1.06864746e+13]


## **np.log():**
#### *Natural logarithm, element-wise.*
El logaritmo natural log es el inverso de la función exponencial, por lo que **log(exp(x)) = x**.  
  El logaritmo natural es logaritmo en base e.


In [1]:
# inspiration and future goal cell

future_job_in_norway = 'Deep Learning Engineer'
print(future_job_in_norway)

Deep Learning Engineer


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
jobs = ['ML & DL Engineer', 'Machine Learning Engineer', 'Deep Learning Engineer']

for fj in jobs:
    print(fj)

ML & DL Engineer
Machine Learning Engineer
Deep Learning Engineer


In [9]:
print('PyTorch = Pythonist')

PyTorch = Pythonist



### Aggregation
#### Aggregation = realizar la misma operación en varias cosas

In [12]:
listyList = [1,2,3]
print('Tipo:', type(listyList))

Tipo: <class 'list'>


In [14]:

print('La suma de la lista: ', sum(listyList))

La suma de la lista:  6


In [15]:
print(a1)

[1 2 3]


#### Cuando hablamos de agregacion, se pueden usar ya sean los metodos de Python, o los de NumPy, ej:

In [16]:
# Metodo de Python:
print(sum(a1))

6


In [18]:
# Metodo de Numpy:
print(np.sum(a1))

6


### Cual es la diferencia, o mejor dicho, que se debe hacer?
#### Se deben usar los metodos de Python para datatypes de Python, y np.methods, para Numpy datatypes, ejemplo sencillo:
*Python datatype:* **`sum()`**  
*NumPy datatype:* **`np.sum()`**


In [20]:
# Un datatype de Python puede ser una lista:
pythonDatatype = [1,2,3,4]
print('Este es un datatype de Python: ', pythonDatatype)
print('La suma de este datatype de Python es:', sum(pythonDatatype), ' con un tipo de dato: ', type(pythonDatatype))

Este es un datatype de Python:  [1, 2, 3, 4]
La suma de este datatype de Python es: 10  con un tipo de dato:  <class 'list'>


In [22]:
# Un datatype de NumPy puede ser un array:
numpyDatatype = np.array([1,2,3,4])
print('Este es un datatype de NumPy: ', numpyDatatype)
print('La suma de este datatype es de: ', np.sum(numpyDatatype), ' con un datatype: ', type(numpyDatatype))

Este es un datatype de NumPy:  [1 2 3 4]
La suma de este datatype es de:  10  con un datatype:  <class 'numpy.ndarray'>


Que NumPy este escrito en ***C***, lo hace increiblemente **`bello`** :D

#### Vamos a crear un array **masivo** para poder ver el poder de NumPy en temas de rendimiento.


In [27]:
massiveArray = np.random.random(100000)
# recordar que np.random.random() -> Return random floats in the half-open interval [0.0, 1.0). 
print('Size: ', massiveArray.size)
print(massiveArray[:100])

Size:  100000
[0.54863815 0.29307534 0.36968783 0.958302   0.3587922  0.75922353
 0.59996502 0.5545326  0.90383358 0.14382119 0.20522983 0.69207692
 0.86390149 0.06245861 0.70423625 0.35672904 0.27699664 0.84602035
 0.2874539  0.2140835  0.04178341 0.35867816 0.84701674 0.24467299
 0.51691255 0.05099817 0.47154585 0.08674462 0.63979932 0.72378207
 0.20975834 0.80286772 0.55933642 0.01419831 0.64457591 0.73563152
 0.98046849 0.74711213 0.58662892 0.65805315 0.12041512 0.30098422
 0.33064067 0.98542152 0.24683758 0.65605787 0.65240322 0.01443469
 0.41371357 0.75588015 0.66381335 0.05089213 0.66320748 0.23414343
 0.5302644  0.97755058 0.35241578 0.78743356 0.70838919 0.57659404
 0.82902651 0.03970263 0.64396332 0.7541073  0.38849046 0.8976632
 0.49526893 0.81156596 0.61529303 0.29601157 0.26550771 0.08299968
 0.74905897 0.98014008 0.11324797 0.24044682 0.61508372 0.97088395
 0.78714364 0.33052401 0.90031019 0.25901389 0.22988247 0.93331258
 0.76537055 0.68458604 0.59386427 0.52747441 0.84

In [29]:
# Lo mismo pero con el nuevo method:
massiveArrayNewMethod = np.random.random_sample(1000)
massiveArrayNewMethod.size
massiveArrayNewMethod[:100]

array([0.98344219, 0.8072781 , 0.67662842, 0.80158276, 0.58321198,
       0.09431391, 0.73278007, 0.362503  , 0.01142723, 0.85939073,
       0.43785127, 0.6605821 , 0.57902524, 0.54340172, 0.45941731,
       0.28541746, 0.33701619, 0.42242173, 0.04690764, 0.88300081,
       0.21746944, 0.77040764, 0.03687579, 0.19823234, 0.02091385,
       0.2826639 , 0.13013647, 0.49871397, 0.61709507, 0.2812661 ,
       0.07646703, 0.61896616, 0.37626225, 0.21273743, 0.89548484,
       0.01333437, 0.82596469, 0.80293574, 0.68209691, 0.35344174,
       0.5325114 , 0.7582954 , 0.06372741, 0.37063255, 0.10808926,
       0.64540107, 0.88262783, 0.73493693, 0.81652341, 0.12919834,
       0.03941271, 0.8080925 , 0.53335105, 0.92656245, 0.91494863,
       0.91444895, 0.89022232, 0.50493401, 0.32580388, 0.29040042,
       0.06530202, 0.90569078, 0.48131998, 0.02037904, 0.56537125,
       0.55641625, 0.09068591, 0.37809366, 0.69521986, 0.61328518,
       0.60176706, 0.51247045, 0.57723346, 0.62189225, 0.20825

### Vamos a comprobar la velocidad de NumPy (escrito en C) vs Python vanilla (por asi decirlo)

In [37]:
# Python (vanilla)
%timeit sum(massiveArray)
# NumPy (escrito en C)
%timeit np.sum(massiveArray)


14.3 ms ± 208 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
49.6 µs ± 1.92 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


Python tarda 14.3 ms, mientras que Nympy tarda 49.6 MICRO segundos, si hacemos calculos:


In [5]:
14300 / 49.6 

288.30645161290323

Lo que nos indica que **NumPy** es **288** veces mas rapido que Python,
que locura, no? :D
Ahora imaginemos esto con cantidades inmensas de datos de una empresa, wow!!

In [40]:
# mas metodos de aggregation:


a2

array([[0. , 1. , 2.3],
       [3. , 4.7, 5. ]])

In [45]:
print(np.mean(a2))
# recordar que np.mean nos retorna el promedio del array
print(np.min(a2))
print(np.max(a2))
print(np.std(a2))
# Devuelve la desviación estándar, una medida de la dispersión de una distribución, de los elementos del array.
# Video del profe Alex de 10 min visto, muy interesante.

2.6666666666666665
0.0
5.0
1.8126101499096698


In [2]:
print(np.var(a2))
# variance = a la medida del grado promedio en el que cada numero es diferente al principal

# higher variance = una gama más amplia de números
#  lower variance = gama inferior de números
# videos de varianza visto, del profe Alex


NameError: name 'np' is not defined

In [4]:
import numpy as np
import pandas as pd


In [8]:
# exercise nut butter store sales


In [11]:
np.random.seed(0)
# number of jars sold
sales_amounts = np.random.randint(20, size=(5,3))

In [15]:
# weekly_sales df with Pandas

weekly_sales = pd.DataFrame(sales_amounts,
                           index=["Monday", "Tuesday", "Wednesday","Thursday", "Friday"],
                           columns=["Almond butter", "Peanut butter", "Cashew butter"])
weekly_sales

Almond butter  Peanut butter  Cashew butter
Monday                12             15              0
Tuesday                3              3              7
Wednesday              9             19             18
Thursday               4              6             12
Friday                 1              6              7

In [20]:
# create butter_prices and DF
prices = np.array([10,8,12])
prices

# el parametro prices nos daria el error: ValueError: Shape of passed values is (3, 1), indices imply (1, 3)
                             # por lo cual debemos usar reshape:
butter_prices = pd.DataFrame(prices.reshape(1,3),
                            index=["Price"],
                            columns=["Almond butter", "Peanut butter", "Cashew butter"])                                
butter_prices

Almond butter  Peanut butter  Cashew butter
Price             10              8             12

In [22]:
total_sales = prices.dot(sales_amounts)
total_sales


ValueError: shapes (3,) and (5,3) not aligned: 3 (dim 0) != 5 (dim 0)

In [24]:
prices.shape, sales_amounts.shape

((3,), (5, 3))

In [27]:
# shapes aren't aligned, let's transpose
total_sales = prices.dot(sales_amounts.T)
total_sales


array([240, 138, 458, 232, 142])

In [29]:
butter_prices.shape, weekly_sales.shape

((1, 3), (5, 3))

In [30]:
# create daily_sales

daily_sales = butter_prices.dot(weekly_sales.T)
daily_sales

Monday  Tuesday  Wednesday  Thursday  Friday
Price     240      138        458       232     142

In [33]:
# does not work, different shapes, let's fixed:
# Output: Expected a 1D array, got an array with shape (5, 5)
weekly_sales["Total $"] = daily_sales
daily_sales

ValueError: Expected a 1D array, got an array with shape (5, 5)

In [1]:
# fixed:
weekly_sales["Total $"] = daily_sales.T
weekly_sales

NameError: name 'daily_sales' is not defined

In [11]:
a1 = np.array([1,2,3])
a1
a2 = np.array([[1,2.2,3.6],
             [4.5,6.5,7.8]])
a2

array([[1. , 2.2, 3.6],
       [4.5, 6.5, 7.8]])

## Operators

In [16]:
a1 > a2
a1 >= a2
a2 >= a1

array([[ True,  True,  True],
       [ True,  True,  True]])

### basicamente los mismos que en Python:  
***https://numpy.org/doc/stable/reference/routines.logic.html***

In [20]:
sort_array = np.array([1,2,3,4,5,6])

for s in sort_array:
    print(s)

1
2
3
4
5
6


## sorting Array in **Numpy**

In [25]:
random_arr = np.random.randint(11, size=(3,5))
random_arr

array([[ 8,  2,  7,  8,  6],
       [10,  0,  2,  1,  9],
       [ 4,  5,  0,  9,  2]])

In [29]:
n1 = np.sort(random_arr)
n1

array([[ 2,  6,  7,  8,  8],
       [ 0,  1,  2,  9, 10],
       [ 0,  2,  4,  5,  9]])

In [31]:
# por index

n2 = np.argsort(random_arr)
n2

array([[0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4]], dtype=int64)

In [35]:
# min index
n2 = np.array([1,2,3,4,5,6])
np.argmin(n2)

#  index max
np.argmax(n2)

5

In [41]:
# con matrices de mas dimensiones ese mas complejo, ojo con esto:

# en axis=0 nos devuelve el maximo pero por columnas, es decir:
print(random_arr)
print("------------------------------")
print(np.argmax(random_arr, axis=0))


[[ 2  6  7  8  8]
 [ 0  1  2  9 10]
 [ 0  2  4  5  9]]
------------------------------
[0 0 0 1 1]


In [44]:
# mientras que xon axis=1, nos devuelve los mayores, pero por filas:
print(random_arr)
print("------------------------------")
print(np.argmax(random_arr, axis=1))

[[ 2  6  7  8  8]
 [ 0  1  2  9 10]
 [ 0  2  4  5  9]]
------------------------------
[3 4 4]


### complejo y demas, pero *hermoso!*

# Turn images into **NumPy Array**, ***NumPy in action!***

<img src="./panda.png"/>